File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/manager/AsyncAssetManager.java
package com.greentree.engine.moon.assets.serializator.manager;

import com.greentree.commons.reflection.info.TypeInfo;
import com.greentree.commons.reflection.info.TypeInfoBuilder;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.key.ResourceAssetKey;
import com.greentree.engine.moon.assets.key.ResultAssetKey;
import com.greentree.engine.moon.assets.serializator.request.KeyLoadRequest;
import com.greentree.engine.moon.assets.serializator.request.builder.KeyRequestBuilder;
import com.greentree.engine.moon.assets.serializator.request.builder.KeyRequestBuilderImpl;
import com.greentree.engine.moon.assets.value.Value;

public interface AsyncAssetManager extends ValueAssetManager {

    default <T> KeyRequestBuilder<T, Value<T>> loadAsync(Class<T> cls) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return loadAsync(type);
    }

    default <T> KeyRequestBuilder<T, Value<T>> loadAsync(TypeInfo<T> type) {
        return new KeyRequestBuilderImpl<>(r -> loadAsync(r), type);
    }

    default <T> Value<T> loadAsync(KeyLoadRequest<T> request) {
        return loadAsync(request.loadType(), request.key(), request.getDefault());
    }

    <T> Value<T> loadAsync(TypeInfo<T> type, AssetKey key, T def);

    default <T> Value<T> loadAsync(Class<T> cls, AssetKey key) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return loadAsync(type, key, loadDefault(type));
    }

    default <T> Value<T> loadAsync(TypeInfo<T> type, AssetKey key) {
        return loadAsync(type, key, loadDefault(type));
    }

    default <T> Value<T> loadAsync(Class<T> cls, String path) {
        return loadAsync(cls, new ResourceAssetKey(path), loadDefault(cls));
    }

    default <T> Value<T> loadAsync(Class<T> cls, AssetKey key, T def) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return loadAsync(type, key, loadDefault(type, def));
    }

    default <T> Value<T> loadAsync(Class<T> cls, String path, T def) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return loadAsync(type, new ResourceAssetKey(path), loadDefault(type, def));
    }

    default <T> Value<T> loadAsync(TypeInfo<T> type, String path) {
        return loadAsync(type, new ResourceAssetKey(path), loadDefault(type));
    }

    default <T> Value<T> loadAsync(TypeInfo<T> type, String path, T def) {
        return loadAsync(type, new ResourceAssetKey(path), def);
    }

    default <T> Value<T> loadAsync(Class<T> cls, Object obj) {
        return loadAsync(cls, new ResultAssetKey(obj), loadDefault(cls));
    }

    default <T> Value<T> loadAsync(Class<T> cls, Object obj, T def) {
        final var type = TypeInfoBuilder.getTypeInfo(cls);
        return loadAsync(type, new ResultAssetKey(obj), loadDefault(type, def));
    }

    default <T> Value<T> loadAsync(TypeInfo<T> type, Object obj) {
        return loadAsync(type, new ResultAssetKey(obj), loadDefault(type));
    }

    default <T> Value<T> loadAsync(TypeInfo<T> type, Object obj, T def) {
        return loadAsync(type, new ResultAssetKey(obj), def);
    }

}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/manager/AsyncAssetManager.kt
package com.greentree.engine.moon.assets.serializator.manager

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.commons.reflection.info.TypeInfoBuilder
import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.asset.ConstAsset
import com.greentree.engine.moon.assets.asset.ReduceAsset
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.AssetKeyType
import com.greentree.engine.moon.assets.key.ResourceAssetKey
import com.greentree.engine.moon.assets.key.ResultAssetKey
import com.greentree.engine.moon.assets.serializator.loader.AssetLoader
import java.lang.Thread.*
import java.util.concurrent.Executors

interface AsyncAssetManager : AssetManager {

	fun <T : Any> load(context: AssetLoader.Context, type: TypeInfo<T>, key: AssetKey): Asset<T>
}

val EXECUTOR = Executors.newFixedThreadPool(
	Runtime.getRuntime().availableProcessors()
) {
	val thread = Thread(it, "async-asset")
	thread.isDaemon = true
	thread.priority = MAX_PRIORITY
	thread
}

fun <T : Any> AsyncAssetManager.loadAsync(type: TypeInfo<T>, key: AssetKey): Asset<T> {
	return object : AssetLoader.Context {
		override fun <T : Any> loadDefault(type: TypeInfo<T>, key: AssetKeyType) =
			this@loadAsync.loadDefault(type, key)

		override fun <T : Any> load(type: TypeInfo<T>, key: AssetKey): Asset<T> {
			val default = loadDefault(type, key.type()) ?: return this@loadAsync.load(this, type, key)
			val result = ReduceAsset(ConstAsset(default))
			EXECUTOR.submit {
				result.asset = this@loadAsync.load(this, type, key)
			}
			return result
		}
	}.load(type, key)
}

fun <T : Any> AsyncAssetManager.loadAsync(cls: Class<T>, key: Any): Asset<T> {
	return loadAsync(cls, ResultAssetKey(key))
}

fun <T : Any> AsyncAssetManager.loadAsync(type: TypeInfo<T>, key: Any): Asset<T> {
	return loadAsync(type, ResultAssetKey(key))
}

fun <T : Any> AsyncAssetManager.loadAsync(cls: Class<T>, key: AssetKey): Asset<T> {
	val type = TypeInfoBuilder.getTypeInfo(cls)
	return loadAsync(type, key)
}

fun <T : Any> AsyncAssetManager.loadAsync(cls: Class<T>, resource: String): Asset<T> {
	return loadAsync(cls, ResourceAssetKey(resource))
}

fun <T : Any> AsyncAssetManager.loadAsync(type: TypeInfo<T>, resource: String): Asset<T> {
	return loadAsync(type, ResourceAssetKey(resource))
}

inline fun <reified T : Any> AsyncAssetManager.loadAsync(key: Any?): Asset<T> {
	return loadAsync(ResultAssetKey(key))
}

inline fun <reified T : Any> AsyncAssetManager.loadAsync(key: AssetKey): Asset<T> {
	val type = TypeInfoBuilder.getTypeInfo(T::class.java)
	return loadAsync(type, key)
}

inline fun <reified T : Any> AsyncAssetManager.loadAsync(resource: String): Asset<T> {
	return loadAsync(ResourceAssetKey(resource))
}